src/mango.ts

```diff
@@ -0,0 +1,493 @@
+    console.log(scalper.symbol, 'Gamma Hedge Too Small');
+    return;
+  }
+
+  const orderIdGamma = new Date().getTime() * 2;
+  const gammaBid = fairValue * (1 - stdDevSpread);
+  const gammaBidID = orderIdGamma + 1;
+  const gammaAsk = fairValue * (1 + stdDevSpread);
+  const gammaAskID = orderIdGamma + 2;
+
+  let gammaFillQty = 0;
+  // TODO: Check fills feed works here
+  fillFeed.removeAllListeners('message');
+  const gammaFillListener = (event) => {
+    const parsedEvent = JSON.parse(event.data);
+    const {
+      owner, clientOrderId, quantity, price,
+    } = parsedEvent.event;
+    if (owner !== mangoAccount.publicKey.toBase58()) {
+      return;
+    }
+    if (clientOrderId === gammaBidID || clientOrderId === gammaAskID) {
+      gammaFillQty += Math.abs(quantity);
+      // Once the gamma fills have crossed the threshold, reset the orders.
+      if (gammaFillQty > netGamma * GAMMA_COMPLETE_THRESHOLD_PCT) {
+        console.log(
+          scalper.symbol,
+          'Gamma Filled',
+          gammaFillQty,
+          clientOrderId === gammaBidID ? 'BOUGHT FOR' : 'SOLD AT',
+          price,
+          new Date().toUTCString(),
+        );
+        fillFeed.removeEventListener('message', gammaFillListener);
+        // Do not need to remove the unfilled order since it will be cancelled in
+        // the recursive call.
+        gammaScalpMango(
+          dipProduct,
+          scalper,
+          mangoClient,
+          mangoAccount,
+          mangoGroup,
+          perpMarket,
+          fillFeed,
+          gammaScalpCount + 1,
+        );
+      } else {
+        console.log('Gamma Partially Filled', gammaFillQty, 'of', netGamma);
+      }
+    }
+  };
+
+  if (fillFeed.readyState === WebSocket.OPEN) {
+    fillFeed.addEventListener('message', gammaFillListener);
+    console.log(scalper.symbol, 'Listening For gamma scalps');
+  } else {
+    console.log(scalper.symbol, 'Websocket State', fillFeed.readyState);
+  }
+
+  // Place gamma scalp bid & offer.
+  try {
+    const gammaBidTx = await mangoClient.perpPlaceOrderIx(
+      mangoGroup,
+      mangoAccount,
+      perpMarket.perpMarketIndex,
+      PerpOrderSide.bid,
+      gammaBid,
+      netGamma,
+      undefined,
+      gammaBidID,
+      PerpOrderType.postOnlySlide,
+    );
+    const gammaAskTx = await mangoClient.perpPlaceOrderIx(
+      mangoGroup,
+      mangoAccount,
+      perpMarket.perpMarketIndex,
+      PerpOrderSide.ask,
+      gammaAsk,
+      netGamma,
+      undefined,
+      gammaBidID,
+      PerpOrderType.postOnlySlide,
+    );
+    const gammaOrdersTx: TransactionInstruction[] = [gammaBidTx, gammaAskTx];
+    console.log(scalper.symbol, 'Gamma Bid', gammaBid, 'ID', gammaBidID);
+    console.log(scalper.symbol, 'Gamma Ask', gammaAsk, 'ID', gammaAskID);
+    await mangoClient.sendAndConfirmTransaction(gammaOrdersTx);
+  } catch (err) {
+    console.log(scalper.symbol, 'Gamma Error', err, err.stack);
```

Comment:
Should this just continue, or would it be best to cleanup and shutdown?
We cant immediately retry since that would get stuck in a feedback loop, but could exponential backoff and retry

Reply:
TODO added. Most common reason for error here will be margin requirements. Which needs closer examination anyway